posts/2023-10-30_floating_point/floating_point.qmd

```diff
@@ -0,0 +1,338 @@
+|----------|----------------------|
+| A \>= B  | A \> B \| near(A, B) |
+| A \<= B  | A \< B \| near(A, B) |
+| A == B   | near(A, B)           |
+| A != B   | !near(A, B)           |
+| A \> B   | A \> B & !near(A, B) |
+| A \< B   | A \< B & !near(A, B) |
+
+
+This would work perfectly fine, but especially for `case_when()` statements, it would add a lot of code-bloat.
+
+
+::: callout-note
+Although a minor issue, it looks like the `near()` function tests for absolute differences, while the `all.equal()` function tests for relative differences, as discussed in [this thread](https://github.com/tidyverse/dplyr/issues/6921):
+
+```{r}
+# very large values:
+# when checking for absolute differences
+near(
+  ANRHI * 1.1 * 10^6,
+  AVAL * 10^6
+)
+# when checking for relative differences
+all.equal(
+  ANRHI * 1.1 * 10^6,
+  AVAL * 10^6
+)
+
+# as:
+(ANRHI * 1.1 * 10^6) %>% format(digits = 22)
+(AVAL * 10^6) %>% format(digits = 22)
+```
+
+:::
+
+A more elegant solution is implemented in the `fpCompare` package, which compares floating point numbers using custom infix operators: 
+
+
+|         `base`            |`fpCompare`    |
+|-------------------------|-------------|
+|        A \>=  B            | A %>=% B	        |
+|        A \<=  B            | A %<=% B	        |
+|        A ==   B           | A %==% B	        |
+|        A !=   B           | A %!=% B	        |
+|        A \>   B            | A %>>% B	        |
+|        A \<   B            | A %<<% B	        |
+
+
+As an example to how this is implemented, we can have a look at the `fpCompare` source code for one of the operators:
+
+```{r}
+`%<=%` <- function(x, y) {
+  (x < y + getOption("fpCompare.tolerance"))
+}
+```
+
+Even if `y` is ever so slightly smaller than `x`, adding the tolerance to `y` will make the result larger than `x`, and the comparison will return `TRUE`.
+
+```{r}
+# we need to set the fpCompare.tolerance first, because we did not load the package:
+options(fpCompare.tolerance = 1e-8)
+
+(ANRHI * 1.1) %<=% AVAL
+```
+
+As long as [{{admiral}}](https://github.com/pharmaverse/admiral) remains open source and free to use, using this package, or even reusing the code itself would be fine. 
+Although this was *my* preferred option, we did not end up implementing it.
+Instead, we made use of the `signif()` function, which rounds a number to a specified number of significant digits.
+This way, we could use the regular infix operators and simply provide the number of significant digits we want to compare to:
+
+```{r}
+signif_dig <- 15
+
+signif(AVAL, signif_dig) == signif(ANRHI * 1.1, signif_dig)
+
+# as:
+(ANRHI * 1.1) %>%
+  signif(signif_dig) %>%
+  format(digits = 22)
+
+# and although when printed, the number still looks off:
+ANRHI <- 101
+((ANRHI * 1.1) %>% signif(signif_dig)) %>% format(digits = 22)
+
+# the comparison works now:
+((ANRHI * 1.1) %>% signif(signif_dig)) == 111.1
+```
+
+This is now implemented throughout `atoxgr_criteria_ctcv5`:
```

Comment:
also throughout `atoxgr_criteria_ctcv4` and `atoxgr_criteria_daids`. We are also looking to implement in `derive_var_anrind` (not yet done).